prow.sh

```diff
@@ -0,0 +1,667 @@
+#   or in its support for different Kubernetes versions (CSIDriverInfo as
+#   CRD in Kubernetes 1.13 vs builtin API in Kubernetes 1.14).
+#
+#   When testing an update for a component in a PR job, the
+#   CSI_PROW_DEPLOYMENT variable can be set in the
+#   .prow.sh of each component when there are breaking changes
+#   that require using a non-default deployment. Otherwise
+#   a deployment named "kubernetes-x.yy" is used.
+#   "none" disables the deployment of the hostpath driver.
+#
+# When no deploy script is found (nothing in `deploy` directory,
+# CSI_PROW_HOSTPATH_REPO=none), nothing gets deployed.
+configvar CSI_PROW_HOSTPATH_VERSION master "hostpath driver"
+configvar CSI_PROW_HOSTPATH_REPO https://github.com/kubernetes-csi/csi-driver-host-path "hostpath repo"
+# Ignore 'See if you can use ${variable//search/replace} instead.'
+# shellcheck disable=SC2001
+configvar CSI_PROW_DEPLOYMENT "kubernetes-$(echo "${CSI_PROW_KUBERNETES_VERSION}" | sed -e 's/\([0-9]*\)\.\([0-9]*\).*/\1.\2/')" "deployment"
+
+# If CSI_PROW_HOSTPATH_CANARY is set (typically to "canary", but also
+# "1.0-canary"), then all image versions are replaced with that
+# version tag.
+configvar CSI_PROW_HOSTPATH_CANARY "" "hostpath image"
+
+# The E2E testing can come from an arbitrary repo. The expectation is that
+# the repo supports "go test ./test/e2e -args --storage.testdriver" (https://github.com/kubernetes/kubernetes/pull/72836)
+# after setting KUBECONFIG. As a special case, if the repository is Kubernetes,
+# then `make WHAT=test/e2e/e2e.test` is called first to ensure that
+# all generated files are present.
+#
+# CSI_PROW_E2E_REPO=none disables E2E testing.
+#
+# TODO (?): support also an E2E test suite in the current repository.
+# TODO: upstream Kubernetes
+configvar CSI_PROW_E2E_REPO https://github.com/pohly/kubernetes "E2E repo"
+configvar CSI_PROW_E2E_VERSION storage-external-snapshot "E2E version"
+configvar CSI_PROW_E2E_IMPORT_PATH k8s.io/kubernetes "E2E package"
+
+# These control which tests run resp. don't run under "ginkgo -p".
+# Tests that depend on alpha features are disabled by default because
+# all tests with a feature tag are skipped.
+#
+# Note that feature tags in the test are set based on what is an alpha
+# feature in the *current* Kubernetes. When updating the test via
+# the CSI_PROW_E2E variables above and testing against an older Kubernetes
+# release, some tests might be enabled by default which don't pass
+# for that older Kubernetes version. In that case, the CSI_PROW_E2E_PARALLEL_SKIP_<k8s-version>
+# variable must be extended. <k8s-version> is just major+minor version
+# separated by underscore.
+configvar CSI_PROW_E2E_PARALLEL_FOCUS 'External.Storage.*csi-hostpath' "parallel E2E focus"
+# Raw block was an alpha feature in 1.13.
+configvar CSI_PROW_E2E_PARALLEL_SKIP_1_13 '|\[Testpattern: Dynamic PV .block volmode.\]' "additional parallel test skip for Kubernetes 1.13"
```

Comment:
If you pull the e2e binary from the 1.13 branch, then you should be able to skip just on the [Feature] tag?

Reply:
True.

But this approach has the potential advantage that new tests can be developed in Kubernetes master and then be applied also when testing against older releases - if it makes sense and the tests pass there, of course.

Reply:
I have wasted many debugging hours due to accidentally running new tests against older releases  that I don't really want to support that here.  We should be getting out of the habit of adding tests post-release.

Reply:
I see what you mean. But the definition of which tests to run is fixed and controlled by the component under testing. So if a PR with a test configuration change shows failures, then that should send a strong message that perhaps that test configuration change is the root cause. It will also be easy to verify whether those failures also occur in other PRs or master.

Regarding "adding tests post-release": consider that a bug is found in a sidecar. We fix that bug and add a test to the E2E suite in master which exposes that bug. Now we backport that sidecar bug fix into an older release which targets a stable Kubernetes release, like 1.13. How do we ensure that the backport actually works if we are limited to running the E2E test suite from 1.13 which doesn't have the new test?

If the bug can be exposed by a unit test in the component or if the component had its own E2E test suite (like the one I'm adding to cluster-driver-registrar in https://github.com/kubernetes-csi/cluster-driver-registrar/pull/41), then we could backport the test together with the fix in a single PR. But that's a lot of ifs, and I am not sure what the Kubernetes release team thinks about backporting tests. Even if they allow it, there is a cyclic dependency: we first have to backport the fix without automatically checking it, do a bugfix release, backport the Kubernetes test such that it uses the new bugfix release, then update the component to use the backported fix. Most likely we'd simply skip testing the fix, or only do it once manually.

Reply:
I still don't understand why specific sidecars would run different K8s e2e tests from each other.  Why can't they just all run "all K8s CSI tests"?  A single sidecar is just one piece in the entire driver and plays one part in the entire K8s volume user experience.  Now, we could potentially try to optimize and say things like "well, external-provisioner only deals with dynamic provisioning, so we only need to run dynamic provisioning tests", but there could end up being functionality that the external-provisioner does that will also impact other functionality not strictly related to dynamic provisioning.  If we filter out specific tests now, then we're not getting full coverage of CSI, and we can easily miss new use cases that get added in the future.

Regarding the scenario of making a bug fix, I think it is reasonable to cherry-pick the new test to 1.13.  There's no great solution to solve the chicken-and-egg problem except to make the fix in the sidecar first, cut an rc, and use the rc in the K8s test.

Reply:
If you want to add tests that don't test the entire driver and only pieces of it (ie just the one sidecar + driver container), then that's more of an integration test rather than a K8s e2e test.

Even if that case, I would not want to encode sidecar-specific test details into a common repository. It should come from configuration stored per-repo.

Reply:
> I still don't understand why specific sidecars would run different K8s e2e tests from each other.

They don't. That's why they'll all share the same config.

But from a design perspective, the configuration is owned by the component, not by the CI. The CI just provides the runtime environment and enough information for the component to identify key differences between different jobs, like Kubernetes version.

The advantage of this approach is that changes to the configuration go through the same process  as code changes, i.e. PR review and testing. Changes on the CI side are minimal and hopefully low risk. For example, switching to a new container image can't break Go compilation because the component itself decides whether it uses the pre-installed Go version.

> If you want to add tests that don't test the entire driver and only pieces of it (ie just the one sidecar + driver container), then that's more of an integration test rather than a K8s e2e test.
>
> Even if that case, I would not want to encode sidecar-specific test details into a common repository. It should come from configuration stored per-repo.

Fully agreed. The actual config in the repo will then have both, the common K8S E2E tests and some additional per-component E2E tests.

Reply:
Let my offer a comparison to illustrate the different approaches: putting configuration into test-infra and rolling it out to all components at once is like shutting down the entire cluster, updating kubelet everywhere and then bring up the cluster again, hoping that everything works.

Putting configuration into `csi-release-tools` and then importing it into each component is like a rolling update of an app on a cluster. If something breaks, stop the update, fix the issue, and try again. Because the update is tested before applying it, there's never any breakage in production. It may be slower to get everything up-to-date, but it's less disruptive and can be automated.